README
Detailed information at: http://www.mediawiki.org/wiki/Extension:Realnames

=Installation=

To install this extension, add the following to LocalSettings.php:
require_once("$IP/extensions/Realnames/Realnames.php");
#add configuration parameters here

=Support=

It's hoped that you will enjoy the extension, and if you do, consider purchasing it as a pay-what-you-want scheme on the author's website ( http://olivierbeaton.com/ ).
If you find a bug or wish to make a feature request, please contact the author by email olivier.beaton@gmail.com. The talk page is also monitored. The author is often in the #mediawiki irc channel under the nick Finlay.


HISTORY
This is the historical changelog for the Realnames extension.

  https://www.mediawiki.org/wiki/Extension:Realnames

See [CHANGELOG](CHANGELOG) for the current releases.

== 0.5.1 (2019-02-04) ==

bugfix:
* added PersonalUrls hook to wfLoadExtension method. Does not work with Timeless skin. (thanks <mhausi>)

== 0.5.0 (2019-02-04) ==

features:
* you can now load extension using wfLoadExtension( 'Realnames' ); 

testing:
* 1.32.0, 1.31.1 (LTS)

== 0.4.0 (2019-02-03) ==

features:
* new $wgRealnamesStyles: dash-reverse, paren-append, paren-reverse
* all $wgRealnamesStyles now include all user text in the link (create custom style to restore previous behaviour)
* $wgRealnamesLinkStyle now defaults to paren-reverse (was reverse)

bugfix:
* fix non-english namespace detection missing colon (:) (thanks Koty, 193.190.131.67)
* fix RecentChanges/AllUsers/etc which added more html to links (thanks Syserr0r)

testing:
* 1.31.1 (LTS)

== 0.3.2 (2019-01-30) ==

notes:
* documentation update

== 0.3.1 (2011-12-25) ==

bugfix:
* custom namespace regex (thanks Cweiske)
* STRICT error with STATIC (thanks Vgold)

notes:
* removed contrib-agreement, consensus in discussions indicate it's
 uneeded to remain BSD in shared environment.
* some progress is being made into core-features for name display, expect
 some future version to support things like sorting based on realnames (thanks Daniel Friesen)
* version 0.3 misreported version as 0.2.1, now 0.3.1 (oops, again)

issues:
* Consensus is that performance could be an issue using current methods. I am currently working
 on a 2-pass version of the extension to address these issues that will be released as soon as
 it is ready.

testing:
* svn trunk 2011-12-26, stable 1.18.0, 1.17.1, 1.16.5, 1.15.5

== 0.3 (2011-11-05) ==

features:
* you can opt-out of replacements using $wgRealnamesReplacements (thanks Salquint)
* we now have smarts in the form of not displaying "Marry - Marry" text
* you can change "smart" features like above using $wgRealnamesSmart

bugfix:
* version 0.2 misreported version as 0.1.1, now 0.3
* now tied into main debugging framework
* a missing return statement was causing titles in append style to get garbled

notes:
* switched to a more permissive and correct bsd-2-clause license agreement for commits
* added more metadata files

issues:
* I continue to be concerned with performance on large recentchanges/history since my initial
 post about it on 2011-09-19. After receiving a few code reviews recently (finally!) the main
 feedback has been that there may indeed be scalability problems with using newFromName. In my
 original post (see extension talk page) I proposed two solutions, but haven't heard from anyone
 experiencing problems yet. I think I may implement a two-pass as described in the next feature
 version, however it will make the code significantly more complex. I'll hold off on extending
 those two specific pages until it's confirmed a two-pass isn't enough.

testing:
* svn trunk 2011-11-05, 1.18.0beta1, 1.17.0, 1.16.5, 1.15.5

== 0.2 (2011-09-22) ==

features:
* replaces username in personnal url bar (where you click login top right vector)
* allows custom namespace detection $wgRealnamesNamespaces
* adds support for Extension:Configure (undoc at release)

bugfix:
* i18n links User: namespaces (thanks Nilsy)
* support for upcoming 1.18 gendered namespaces
* title handling for pre 1.16 wiki's
* custom namespace names

testing:
* trunk 2011-09-22 stable 1.17.0, 1.16.0, 1.15.5

== 0.1.1 (2011-09-17) ==

bugfix:
* eliminated use of create_function due speed implications (thanks Dantman)
* changed user caching to realname cache for a smaller memory footprint (thanks Dantman)
* moved debug handling into class and uses $wgRequest, can expand later for 1.18+ (thanks Dantman)
* license change to BSD, cc is not recommended for software (says them) (thanks Dantman)
* fixed author link (backported to 0.1, thanks Dantman)

issues:
* It may be that this extension is slow on RecentChanges and History pages. If so,
  I have a solution, but I'd rather not introduce more code if I don't have to.
  If you experience noticable slowdowns on these pages please let me know at
  olivier dot beaton at gmail dot com

== 0.1 (2011-09-17) ==

initial tested pages:
* Special:ListUsers
* old and new Special:RecentChanges
* Special:Search of the User: space
* History of page
* Revisions list of page
* Revision view of page
* Revision Compare
* Page header + html title (limited)
* Should integrate into any page where user links are found

issues:
* Always destructive of User: text, new config in future release?
* improve bare detection on users with spaces or underscores in names (article titles and html title)


Realnames.body.php
<?php

/*
	Copyright 2011-2019 Finlay Beaton. All rights reserved.

	Redistribution and use in source and binary forms, with or without modification, are
	permitted provided that the following conditions are met:

		1. Redistributions of source code must retain the above copyright notice, this list of
			conditions and the following disclaimer.

		2. Redistributions in binary form must reproduce the above copyright notice, this list
			of conditions and the following disclaimer in the documentation and/or other materials
			provided with the distribution.

	THIS SOFTWARE IS PROVIDED BY Finlay Beaton ''AS IS'' AND ANY EXPRESS OR IMPLIED
	WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
	FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Finlay Beaton OR
	CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
	CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
	SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
	ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
	NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
	ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

if ( defined( 'MEDIAWIKI' ) === false ) {
	die( 'This file is a MediaWiki extension, it is not a valid entry point' );
}

/**
 * >= 0.1
 */
class ExtRealnames {

	/**
   * A cache of realnames for given users.
	 *
	 * @var   \array
   * @since 2011-09-16, 0.1
   */
	protected static $realnames = array();

	/**
   * namespace regex option string.
	 *
	 * @var   string|null
   * @since 2011-09-16, 0.2
   */
	protected static $namespacePrefixes = null;
	
	/**
   * namespace regex option string urlencoded.
	 *
	 * @var   string|null
   * @since 2019-03-10, 0.6
   */
  protected static $namespacePrefixesEncoded = null;

	/**
   * checks a data set to see if we should proceed with the replacement.
	 *
   * @param \array $matches keyed with regex matches
	 *
   * @return \string text to replace the match with
	 *
   * @since 2011-09-16, 0.1
   * @see   lookForBare() for regex
   */
  protected static function checkBare( $matches ) {
		// matches come from self::lookForBare()'s regular experession
		$m = array(
			'all' => $matches[0],
			'username' => $matches[1],
		);

		self::debug( __METHOD__, print_r( $m, true ) );

		// we do not currently do any checks on Bare replacements, a User: find is
		// always valid but we could add one in the future, and the debug
		// information is still conveniant and keeps things consistent with checkLink
		return self::replace( $m );
  }

	/**
   * checks a data set to see if we should proceed with the replacement.
	 *
   * @param \array $matches keyed with regex matches
	 *
   * @return \string text to replace the match with
	 *
   * @since 2011-09-16, 0.1
   * @see   lookForBare() for regex
   */
  protected static function checkLink( $matches ) {
		// matches come from self::lookForLinks()'s regular experession
		$m = array(
			'all' => $matches[0],
			'linkstart' => $matches[1],
			'linkuser' => $matches[2],
			'username' => $matches[3],
			'linkend' => $matches[4],
		);

		self::debug( __METHOD__, print_r( $m, true ) );

		// some links point to user pages but do not display the username,
		// we can safely ignore those
		// we need to urldecode the link for accents and special characters,
		// and ensure our username has underscores instead of spaces to match our link
		// before being able to do the comparison.
		if ( urldecode( $m['linkuser'] ) !== str_replace( ' ', '_', $m['username'] ) ) {
			return $m['all'];
		}

		return self::replace( $m );
  }

	/**
	 * Outputs to the debug channel.
	 *
	 * @param string $method method name, usually __METHOD__
	 * @param string $text text to log
	 *
	 * @return void
	 *
	 * @since 2019-01-24, 0.3.2
	 */
  protected static function debug( $method, $text ) {
		wfDebugLog( 'realnames', $method . ': ' . $text );
  }

	/**
   * formats the final string in the configured style to display the real name.
	 *
   * @param \array $m keyed with strings called
   *    \li<em>linkstart</em>
   *    \li<em>username</em>
   *    \li<em>realname</em>
   *    \li<em>linkend</em>
	 *
   * @return \string formatted text to replace the match with
	 *
   * @since 2011-09-16, 0.1
   * @see   $wgRealnamesLinkStyle
   * @see   $wgRealnamesBareStyle
   * @see   $wgRealnamesStyles
   * @see   $wgRealnamesBlank
   */
  protected static function display( $m ) {
		// what kind of formatting will we do?
		$style = $GLOBALS['wgRealnamesLinkStyle'];
		$styleBlankName = $GLOBALS['wgRealnamesLinkStyleBlankName'];
		$styleSameName = $GLOBALS['wgRealnamesLinkStyleSameName'];
		if ( empty( $m['linkstart'] ) === true ) {
			if ( $GLOBALS['wgRealnamesBareStyle'] !== false ) {
				$style = $GLOBALS['wgRealnamesBareStyle'];
			}
			if ( $GLOBALS['wgRealnamesBareStyleBlankName'] !== false ) {
				$styleBlankName = $GLOBALS['wgRealnamesBareStyleBlankName'];
			}
			if ( $GLOBALS['wgRealnamesBareStyleSameName'] !== false ) {
				$styleSameName = $GLOBALS['wgRealnamesBareStyleSameName'];
			}
			$m['linkstart'] = '';
			$m['linkend'] = '';
		}

		if ( empty( $style ) === true ) {
			// error
			self::debug( __METHOD__, 'error, blank style configuration' );
			return $m['all'];
		}

		// get the formatting code
		$format = $GLOBALS['wgRealnamesStyles'][$style];

		if ( empty( $style ) === true ) {
			// error
			self::debug( __METHOD__, 'error, blank format configuration' );
			return $m['all'];
		}

		// we have a blank realname, and the admin doesn't want to see them,
		// or his chosen format will not display a username at all
		if ( empty( $m['realname'] ) === true && (
			$GLOBALS['wgRealnamesBlank'] === false || strpos( $format, '$2' ) === false
			) ) {
			$format = $GLOBALS['wgRealnamesStyles'][$styleBlankName];
		}

		if ( $GLOBALS['wgRealnamesSmart'] !== false
			&& $GLOBALS['wgRealnamesSmart']['same'] === true
			&& $m['username'] === $m['realname']
			&& strpos( $format, '$2' ) !== false
			&& strpos( $format, '$3' ) !== false
			) {
			// we only do this if both username and realname will be displayed in
			// the user's format
			self::debug( __METHOD__, 'smart dupe detected' );

			// we're going to display: John - John
			// this is silly. The smart thing to do
			// is infact nothing (in the name)
			$format = $GLOBALS['wgRealnamesStyles'][$styleSameName];
		}

		// plug in our values to the format desired
		// redo to ensure order
		$text = wfMsgReplaceArgs( $format, array(
			$m['linkstart'],
			str_replace( '_', ' ', $m['username'] ),
			str_replace( '_', ' ', $m['realname'] ),
			$m['linkend'],
			) );

		self::debug( __METHOD__, 'replacing with ' . print_r( $text, true ) );

		return $text;
  }

	/**
   * gather list of namespace prefixes in the wiki's language.
   * this is a regex string.
	 *
   * @return \string regex namespace options
	 *
   * @since 2011-09-22, 0.2
   */
  public static function getNamespacePrefixes($encode = false) {
		if ($encode === true) {
			$prefixes = self::$namespacePrefixesEncoded;			
		} else {
			$prefixes = self::$namespacePrefixes;
		}

		// if we already figured it all out, just use that again
		if ( $prefixes !== null ) {
			return $prefixes;
		}

		// always catch this one
		$namespaces = array(
			'User',
			'User_talk',
			'User talk',
		);

		// add in user specified ones
		$namespaces = array_merge( $namespaces, array_values( $GLOBALS['wgRealnamesNamespaces'] ) );

		// try to figure out the wiki language
		// ! get language from the context somehow? (2011-09-26, ofb)
		$lang = $GLOBALS['wgContLang'];

		// user namespace's primary name in the wiki lang
		$namespaces[] = $lang->getNsText( NS_USER );
		$namespaces[] = $lang->getNsText( NS_USER_TALK );

		// namespace aliases and gendered namespaces (1.18+) in the wiki's lang
		// fallback for pre 1.16
		if ( method_exists( $lang, 'getNamespaceAliases' ) === true ) {
			$nss = $lang->getNamespaceAliases();
		} else {
			$nss = $GLOBALS['wgNamespaceAliases'];
		}

		foreach ( $nss as $name => $space ) {
			if ( in_array( $space, array( NS_USER, NS_USER_TALK ) ) === true ) {
				$namespaces[] = $name;
			}
		}

		// clean up
		$namespaces = array_unique( $namespaces );

		if ($encode === true) {
			$namespaces = array_map('urlencode', $namespaces);
		}

		$prefixes = '(?:(?:' . implode( '|', $namespaces ) . '):)';

		self::debug( __METHOD__, 'namespace prefixes: ' . $prefixes );

		if ($encode === true) {
			self::$namespacePrefixesEncoded = $prefixes;			
		} else {
			self::$namespacePrefixes = $prefixes;
		}

		return $prefixes;
  }

	/**
   * >= 0.1, change all usernames to realnames.
	 *
   * @param OutputPage &$out The OutputPage object.
   * @param Skin &$skin object that will be used to generate the page, added in 1.13.
   *
   * @return \bool true, continue hook processing
	 *
   * @since 2011-09-16, 0.1
   * @note  OutputPageBeforeHTML does not work for Special pages like RecentChanges or ActiveUsers
	 * @note  requires MediaWiki 1.7.0
	 * @see   hook documentation http://www.mediawiki.org/wiki/Manual:Hooks/BeforePageDisplay
   */
  public static function hookBeforePageDisplay( &$out, &$skin = false ) {
		// pre 1.16 no getTitle()
		if ( method_exists( $out, 'getTitle' ) === true ) {
			$title = $out->getTitle();
		} else {
			$title = $GLOBALS['wgTitle'];
		}

		if ( $GLOBALS['wgRealnamesReplacements']['title'] === true ) {
			// article title
			self::debug( __METHOD__, 'searching article title...' );

			// special user page handling
			// User:
			if ( in_array( $title->getNamespace(), array( NS_USER, NS_USER_TALK ) ) === true ) {
				// swap out the specific username from title
				// this overcomes the problem lookForBare has with spaces and underscores in names
				$reg = '/'
					. self::getNamespacePrefixes()
					. '\s*('
					. $title->getText()
					. ')(?:\/.+)?/';
				$bare = self::lookForBare(
					$out->getPageTitle(),
					$reg
				);
				$out->setPagetitle( $bare );
			}

			// this should also affect the html head title
			$out->setPageTitle( self::lookForBare( $out->getPageTitle() ) );
		}

		if ( $GLOBALS['wgRealnamesReplacements']['subtitle'] === true ) {
			// subtitle (say, on revision pages)
			self::debug( __METHOD__, 'searching article subtitle...' );
			$out->setSubtitle( self::lookForLinks( $out->getSubtitle() ) );
		}

		if ( $GLOBALS['wgRealnamesReplacements']['body'] === true ) {
			// article html text
			self::debug( __METHOD__, 'searching article body...' );
			$out->mBodytext = self::lookForLinks( $out->getHTML() );
		}

		return true;
  }

  /**
	 * >= 0.2, change all usernames to realnames in url bar.
   * change all usernames to realnames in skin top right links bar
	 *
	 * @param \array &$personal_urls the array of URLs set up so far
   * @param Title $title the Title object of the current article
	 *
   * @return \bool true, continue hook processing
	 *
	 * @since 2011-09-22, 0.2
   * @see   hook documentation http://www.mediawiki.org/wiki/Manual:Hooks/PersonalUrls
   * @note  requires MediaWiki 1.7.0
   * @note  does nothing for Timeless skin
   */
  public static function hookPersonalUrls( &$personal_urls, $title ) {
		if ( $GLOBALS['wgRealnamesReplacements']['personnal'] === true ) {
			self::debug( __METHOD__, 'searching personnal urls...' );

			// replace the name of the logged in user
			if ( isset( $personal_urls['userpage'] ) === true
				&& isset( $personal_urls['userpage']['text'] ) === true ) {
				// fake the match, we know it's there
				$m = array(
					'all' => $personal_urls['userpage']['text'],
					'username' => $personal_urls['userpage']['text'],
					'realname' => $GLOBALS['wgUser']->getRealname(),
					);
				$personal_urls['userpage']['text'] = self::replace( $m );
			}
		}

		return true;
  }

  /**
   * scan and replace plain usernames of the form User:username into real names.
	 *
   * @param \string $text to scan
   * @param \string $pattern to match, \bool false for default
	 *
   * @return \string with realnames replaced in
	 *
   * @since 2011-09-16, 0.1
   * @note  bug: we have problems with users with underscores (they become spaces) or spaces,
   *    we tend to just strip the User: and leave the username, but we only modify the
   *    first word so some weird style might screw it up (2011-09-17, ofb)
   */
  protected static function lookForBare( $text, $pattern = false ) {
		if ( empty( $pattern ) === true ) {
			// considered doing [^<]+ here to catch names with spaces or underscores,
			// which works for most titles but is not universal
			$pattern = '/' . self::getNamespacePrefixes() . '([^ \t]+)(\/.+)?/';
		}

		self::debug( __METHOD__, 'pattern: ' . $pattern );
		// create_function is slow
		$ret = preg_replace_callback(
			$pattern,
			array(
				__CLASS__,
				'checkBare',
			),
			$text
		);

		return $ret;
	}

	/**
   * scan and replace username links into realname links.
	 *
   * @param \string $text to scan
   * @param \string $pattern to match, \bool false for default
	 *
   * @return \string with realnames replaced in
	 *
   * @since 2011-09-16, 0.1
   */
	protected static function lookForLinks( $text, $pattern = false ) {
		self::debug(__METHOD__, 'before: '.self::getNamespacePrefixes(false));
		self::debug(__METHOD__, 'after: '.self::getNamespacePrefixes(true));
		if ( empty( $pattern ) === true ) {
			$pattern = '/(<a\b[^">]+href="[^">]+'
				. self::getNamespacePrefixes(true)
				. '([^"\\?\\&>]+)[^>]+>(?:<bdi>)?)'
				. self::getNamespacePrefixes()
				. '?([^>]+)((?:<\\/bdi>)?<\\/a>)/';
		}

		// create_function is slow
		$preg = preg_replace_callback(
			$pattern,
			array(
				__CLASS__,
				'checkLink',
			),
			$text
		);
		return $preg;
	}

	/**
   * obtains user information based on a match for future replacement.
   *
	 * @param \array $m keyed with strings called
   *    \li<em>linkstart</em> (optional)
   *    \li<em>username</em>
   *    \li<em>realname</em> (optional)
   *    \li<em>linkend</em> (optional)
	 *
   * @return \string formatted text to replace the match with
	 *
   * @since 2011-09-16, 0.1
   */
	protected static function replace( $m ) {
		$debug_msg = 'matched '
			. ( ( isset( $m['username'] ) === true ) ? $m['username'] : print_r( $m, true ) );
		self::debug( __METHOD__, $debug_msg );

		if ( isset( self::$realnames[$m['username']] ) === false ) {
			// we don't have it cached
			$realname = null;

			if ( isset( $m['realname'] ) === true ) {
				// we got it elsewhere
				$realname = $m['realname'];
			} else {
				// time to do a lookup
				$user = User::newFromName( $m['username'] );

				if ( is_object( $user ) === false ) {
					self::debug( __METHOD__, 'skipped, invalid user: ' . $m['username'] );
					return $m['all'];
				}

				$realname = $user->getRealname();
			}

			self::$realnames[$m['username']] = htmlspecialchars( trim( $realname ) );
		}

		// this may be blank
		$m['realname'] = self::$realnames[$m['username']];

		return self::display( $m );
	}
}
